File: dns.go
package main

import (
	"fmt"
	"os"
	"time"

	container "github.com/scaleway/scaleway-sdk-go/api/container/v1beta1"
	domain "github.com/scaleway/scaleway-sdk-go/api/domain/v2beta1"
	"github.com/scaleway/scaleway-sdk-go/scw"
)

const (
	waitForDNSDefaultTimeout = 15 * time.Minute
	defaultRetryInterval     = 5 * time.Second
)

const (
	CNAME = "CNAME"
	ALIAS = "ALIAS"
)

var (
	TTL = uint32(360)
)

func WaitForDNSReady(client *scw.Client, DNSZone string) (*domain.DNSZone, error) {
	fmt.Println("waiting for dns zone to be ready")

	api := domain.NewAPI(client)

	dns, err := api.WaitForDNSZone(&domain.WaitForDNSZoneRequest{
		DNSZone: DNSZone,
	})

	if err != nil {
		return dns, err
	}

	return dns, nil

}

func DeleteDNSRecord(
	client *scw.Client,
	Container *container.Container,
	DNSZone string,
) (*domain.UpdateDNSZoneRecordsResponse, error) {
	fmt.Println("Update Zone DNS - Delete")

	// ENV
	Prefix := os.Getenv(EnvDNSPrefix)
	RootZone := os.Getenv(EnvRootZone)

	api := domain.NewAPI(client)

	Data := Container.DomainName + "."

	var Name string = Container.Name
	var Type domain.RecordType = CNAME

	// Handle Prefix DNS

	if Prefix != "" {
		Name = Prefix

		fmt.Println("Update With Prefix Zone DNS - Delete", Prefix)
	}

	// Handle Root Zone Alias
	// Some DNS doesn't handle correctly CNAME on Root Zone.
	// We should use an Alias

	if RootZone == "true" {
		Name = ""
		Type = ALIAS
		fmt.Println("Update Root Zone DNS - Delete")
	}

	IDFields := &domain.RecordIdentifier{
		Name: Name,
		Data: &Data,
		Type: Type,
		TTL:  &TTL,
	}

	Changes := []*domain.RecordChange{
		{
			Delete: &domain.RecordChangeDelete{
				IDFields: IDFields,
			},
		},
	}

	records, err := api.UpdateDNSZoneRecords(
		&domain.UpdateDNSZoneRecordsRequest{
			DNSZone: DNSZone,
			Changes: Changes,
		})

	if err != nil {

		return nil, err
	}

	return records, nil
}

func SetDNSRecord(
	client *scw.Client,
	Container *container.Container,
	DNSZone string,
) (string, error) {

	// ENV
	Prefix := os.Getenv(EnvDNSPrefix)
	RootZone := os.Getenv(EnvRootZone)

	fmt.Println("Update Zone DNS - Add")

	api := domain.NewAPI(client)

	var Name string = Container.Name
	var Type domain.RecordType = CNAME

	// Handle Prefix DNS

	if Prefix != "" {
		Name = Prefix

		fmt.Println("Update With Prefix Zone DNS - Add", Prefix)
	}

	var Hostname string = Name + "." + DNSZone

	// Handle Root Zone Alias
	// Some DNS doesn't handle correctly CNAME on Root Zone.
	// We should use an Alias

	if RootZone == "true" {
		Name = ""
		Type = ALIAS
		Hostname = DNSZone
		fmt.Println("Update Root Zone DNS - Add")
	}

	Records := []*domain.Record{
		{
			Name: Name,
			Type: Type,
			TTL:  TTL,
			Data: Container.DomainName + ".",
		},
	}

	Data :=  Container.DomainName + "."

	IDFields := &domain.RecordIdentifier{
		Name: Name,
		Type: Type,
		TTL:  &TTL,
		Data: &Data,
	}

	Changes := []*domain.RecordChange{
		{
			Set: &domain.RecordChangeSet{
				IDFields: IDFields,
				Records: Records,
			},
		},
	}

	_, err := api.UpdateDNSZoneRecords(
		&domain.UpdateDNSZoneRecordsRequest{
			DNSZone: DNSZone,
			Changes: Changes,
		})

	if err != nil {
		return "", err
	}

	fmt.Println("Hostname", Hostname)

	return Hostname,nil
}

File: src/dns.ts
import * as core from '@actions/core';
import { Container as ContainerSDK, Domain as DomainSDK, Client } from '@scaleway/sdk';
import { ENV, DNS } from './constants';

export async function deleteDnsRecord(
  client: Client,
  container: any,
  dnsZone: string
): Promise<any> {
  core.info('Update Zone DNS - Delete');
  
  const prefix = process.env[ENV.DNS_PREFIX] || '';
  const rootZone = process.env[ENV.ROOT_ZONE] || 'false';
  
  const api = new DomainSDK.v2beta1.API(client);
  
  const data = `${container.domainName}.`;
  
  let name = container.name;
  let type = DNS.CNAME;
  
  if (prefix) {
    name = prefix;
    core.info(`Update With Prefix Zone DNS - Delete: ${prefix}`);
  }
  
  if (rootZone === 'true') {
    name = '';
    type = 'ALIAS' as any;
    core.info('Update Root Zone DNS - Delete');
  }
  
  const changes = [
    {
      delete: {
        idFields: {
          name,
          data,
          type,
          ttl: DNS.TTL,
        },
      },
    },
  ];
  
  const response = await api.updateDNSZoneRecords({
    dnsZone,
    changes,
  } as any);
  
  return response;
}

export async function setDnsRecord(
  client: Client,
  container: any,
  dnsZone: string
): Promise<string> {
  const prefix = process.env[ENV.DNS_PREFIX] || '';
  const rootZone = process.env[ENV.ROOT_ZONE] || 'false';
  
  core.info('Update Zone DNS - Add');
  
  const api = new DomainSDK.v2beta1.API(client);
  
  let name = container.name;
  let type = DNS.CNAME;
  
  if (prefix) {
    name = prefix;
    core.info(`Update With Prefix Zone DNS - Add: ${prefix}`);
  }
  
  let hostname = `${name}.${dnsZone}`;
  
  if (rootZone === 'true') {
    name = '';
    type = 'ALIAS' as any;
    hostname = dnsZone;
    core.info('Update Root Zone DNS - Add');
  }
  
  const records = [
    {
      name,
      type,
      ttl: DNS.TTL,
      data: `${container.domainName}.`,
    },
  ];
  
  const data = `${container.domainName}.`;
  
  const changes = [
    {
      set: {
        idFields: {
          name,
          type,
          ttl: DNS.TTL,
          data,
        },
        records,
      },
    },
  ];
  
  await api.updateDNSZoneRecords({
    dnsZone,
    changes,
  } as any);
  
  core.info(`Hostname: ${hostname}`);
  
  return hostname;
}